Amazon.IonObjectMapper/IonObjectSerializer.cs

```diff
@@ -175,6 +164,150 @@ public override void Serialize(IIonWriter writer, object item)
+        {
+            // We deserialize whether or not we ultimately use the result because values
+            // for some Ion types need to be consumed in order to advance the reader.
+            deserialized = ionSerializer.Deserialize(reader, field.FieldType, ionType);
+            
+            if (options.IgnoreReadOnlyFields && field.IsInitOnly)
+            {
+                return false;
+            }
+
+            return !options.IgnoreDefaults || deserialized != default;
+        }
+
+        private object DeserializeWithIonConstructor(ConstructorInfo ionConstructor, IIonReader reader)
+        {
+            var parameters = ionConstructor.GetParameters();
+
+            // Compute mapping between parameter names and index in parameter array so we can figure out the
+            // correct order of the constructor arguments.
+            var paramIndexMap = new Dictionary<string, int>();
+            for (int i = 0; i < parameters.Length; i++)
+            {
+                var ionPropertyName = (IonPropertyName) parameters[i].GetCustomAttribute(typeof(IonPropertyName));
+                if (ionPropertyName == null)
+                {
+                    throw new NotSupportedException(
+                        $"Parameter '{parameters[i].Name}' is not specified with the [IonPropertyName] attribute " +
+                        $"for {targetType.Name}'s IonConstructor. All constructor arguments must be annotated " +
+                        "so we know which parameters to set at construction time.");
+                }
+
+                paramIndexMap.Add(ionPropertyName.Name, i);
+            }
+
+            reader.StepIn();
+
+            // Deserialize Ion and organize deserialized results into four categories:
+            // 1. Values to be passed into the Ion constructor.
+            // 2. Values to be set via annotated methods after construction.
+            // 3. Properties to be set after construction.
+            // 4. Fields to be set after construction.
+            var constructorArgs = new object[parameters.Length];
+            var setterMethods = new List<(MethodInfo, object)>();
+            var remainingProperties = new List<(PropertyInfo, object)>();
+            var remainingFields = new List<(FieldInfo, object)>();
+            IonType ionType;
+            while ((ionType = reader.MoveNext()) != IonType.None)
+            {
+                MethodInfo method;
+                PropertyInfo property;
+                FieldInfo field;
+                if (paramIndexMap.ContainsKey(reader.CurrentFieldName))
+                {
+                    var index = paramIndexMap[reader.CurrentFieldName];
+                    var deserialized = ionSerializer.Deserialize(reader, parameters[index].ParameterType, ionType);
+                    constructorArgs[index] = deserialized;
+                }
+                else if ((method = FindSetter(reader.CurrentFieldName)) != null)
```

Comment:
Specifically, these are the duplicated parts, which we need to try to avoid.

Reply:
The two implementations have merged into one so there is no more duplication.